aptos-move/framework/aptos-framework/sources/scheduled_transactions/scheduled_txns.move

```
@@ -0,0 +1,1171 @@
+            scheduled_time_ms,
+            max_gas_amount,
+            max_gas_unit_price,
+            pass_signer,
+            f: ScheduledFunction::V1(f)
+        }
+    }
+
+    /// Insert a scheduled transaction into the queue. ScheduleMapKey is returned to user, which can be used to cancel the txn.
+    public fun insert(
+        sender: &signer, txn: ScheduledTransaction
+    ): ScheduleMapKey acquires ScheduleQueue, AuxiliaryData {
+        // If scheduling is shutdown, we cannot schedule any more transactions
+        let aux_data = borrow_global<AuxiliaryData>(@aptos_framework);
+        assert!(
+            (aux_data.module_status == ScheduledTxnsModuleStatus::Active),
+            error::unavailable(EUNAVAILABLE)
+        );
+
+        // we expect the sender to be a permissioned signer
+        assert!(
+            signer::address_of(sender) == txn.sender_addr,
+            error::permission_denied(EINVALID_SIGNER)
+        );
+
+        // Only schedule txns in the future
+        let txn_time = txn.scheduled_time_ms;
+        let block_time = timestamp::now_microseconds() / 1000;
+        assert!(txn_time > block_time, error::invalid_argument(EINVALID_TIME));
+
+        assert!(
+            txn.max_gas_unit_price >= 100,
+            error::invalid_argument(ELOW_GAS_UNIT_PRICE)
+        );
+
+        assert!(
+            bcs::serialized_size(&txn) < MAX_SCHED_TXN_SIZE,
+            error::invalid_argument(ETXN_TOO_LARGE)
+        );
+
+        // Generate unique transaction ID
+        let hash = sha3_256(bcs::to_bytes(&txn));
+        let txn_id = hash_to_u256(hash);
+
+        // Insert the transaction into the schedule_map
+        // Create schedule map key
+        let key = ScheduleMapKey {
+            time: txn_time,
+            gas_priority: U64_MAX - txn.max_gas_unit_price,
+            txn_id
+        };
+
+        let queue = borrow_global_mut<ScheduleQueue>(@aptos_framework);
+        queue.schedule_map.add(key, Empty {});
```

if the txn is scheduled for multiple times, seems we are going to abort, should we return a failure instead?

---

Sure. But will do it once the feature is more stable, probably before the final draft